src/ivectorbin/ivector-extract-dense.cc

```diff
@@ -0,0 +1,313 @@
+    sequencer_config.Register(&po);
+
+    po.Read(argc, argv);
+
+    if (po.NumArgs() != 5 && po.NumArgs() != 7) {
+      po.PrintUsage();
+      exit(1);
+    }
+
+    std::string ivector_extractor_rxfilename = po.GetArg(1),
+      feature_rspecifier = po.GetArg(2),
+      posterior_rspecifier = po.GetArg(3);
+    IvectorExtractor extractor;
+    ReadKaldiObject(ivector_extractor_rxfilename, &extractor);
+    RandomAccessPosteriorReader posterior_reader(posterior_rspecifier);
+    double tot_auxf_change = 0.0, tot_t = 0.0;
+    double *auxf_ptr = (compute_objf_change ? &tot_auxf_change : NULL );
+    int32 num_done = 0, num_err = 0;
+
+    if (po.NumArgs() == 5) {
+      std::string ivector_wspecifier = po.GetArg(4),
+        utt2spk_wspecifier = po.GetArg(5);
+      SequentialBaseFloatMatrixReader feature_reader(feature_rspecifier);
+      BaseFloatVectorWriter ivector_writer(ivector_wspecifier);
+      TokenWriter utt2spk_writer(utt2spk_wspecifier);
+      for (; !feature_reader.Done(); feature_reader.Next()) {
+        std::string utt = feature_reader.Key();
+        const Matrix<BaseFloat> &feat = feature_reader.Value();
+        Posterior post = posterior_reader.Value(utt);
+        if (static_cast<int32>(post.size()) != feat.NumRows()) {
+          KALDI_WARN << "Size mismatch between posterior " << post.size()
+            << " and features " << feat.NumRows() << " for utterance " << utt;
+          num_err++;
+          continue;
+        }
+        double this_t = opts.acoustic_weight * TotalPosterior(post),
+          max_count_scale = 1.0;
+        if (opts.max_count > 0 && this_t > opts.max_count) {
+          max_count_scale = opts.max_count / this_t;
+          KALDI_LOG << "Scaling stats for utterance " << utt
+            << " by scale " << max_count_scale << " due to --max-count="
+            << opts.max_count;
+          this_t = opts.max_count;
+        }
+        ScalePosterior(opts.acoustic_weight * max_count_scale,
+                       &post);
+        int32 num_chunks = NumChunks(feat.NumRows(), chunk_size, period);
+        for (int32 i = 0; i < num_chunks; i++) {
+          Vector<BaseFloat> ivector(extractor.IvectorDim());
+          int32 chunk_start, chunk_end;
+          GetChunkRange(feat.NumRows(), i, chunk_size, period, min_chunk,
+            &chunk_start, &chunk_end);
+          IvectorExtractFromChunk(extractor, utt, feat, post, chunk_start,
+            chunk_end, &ivector, auxf_ptr);
+          std::stringstream ss_segment;
+          ss_segment << utt << "-"
+            << std::setw(6) << std::setfill('0') << chunk_start << std::setw(1)
+            << "-" << std::setw(6) << std::setfill('0') << chunk_end;
+          std::string segment = ss_segment.str();
+          ivector_writer.Write(segment, ivector);
+          utt2spk_writer.Write(segment, utt);
+          tot_t += this_t;
+          num_done++;
+        }
+      }
+    } else {
+      std::string segments_rxfilename = po.GetArg(4),
```

Comment:
this main() is kind of long.  I wonder if it would make it clearer if some parts of it were broken out somehow... think about it.

Reply:
There is no necessity for this binary. There is a script https://github.com/kaldi-asr/kaldi/blob/master/egs/wsj/s5/utils/data/get_uniform_subsegments.py to create uniform segments and this can be passed to the normal ivector-extract function.

Reply:
@danpovey I am going to start by focusing on seeing if we can use the script Vimal recommends and eliminate the ivector-extract-dense binary.